walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,157 @@
+   * @param addedTimes All the {@code TimeRange} which do not overlap
+   * @param time A time to be added to {@code addedTimes}. {@code addedTimes} is changed to include time
+   * without overlapping any {@code TimeRange}
+   */
+  private void mergeToAddedTimes(Collection<TimeRange> addedTimes, TimeRange time) {
+    Collection<TimeRange> overlappingTimes = new ArrayList<TimeRange>();
+    // find all overlapping times
+    for (TimeRange includedTime : addedTimes) {
+      if (includedTime.overlaps(time)) {
+        overlappingTimes.add(includedTime);
+      }
+    }
+
+    int minStart = time.start();
+    int maxEnd = time.end();
+    for (TimeRange t : overlappingTimes) {
+      if (t.start() < minStart) minStart = t.start();
+      if (t.end() > maxEnd) maxEnd = t.end();
+      addedTimes.remove(t);
+    }
+    addedTimes.add(TimeRange.fromStartEnd(minStart, maxEnd, false));
+  }
+
+  /**
+   * From a potentially overlapping collection of times, this method finds all the time ranges covered
+   * @param times A potentially overlapping collection of {@code TimeRange}
+   * @return A collection of {@code TimeRange} which covers the same times as {@code times}, but has no
+   * overlapping {@code TimeRange} 
+   */
+  private Collection<TimeRange> getNonOverlappingTimes(Collection<TimeRange> times) {
+    Collection<TimeRange> timeRanges = new ArrayList<TimeRange>();
+    for (TimeRange t : times)
+      mergeToAddedTimes(timeRanges, t);
+    return timeRanges;
+  }
+
+  /**
+   * Get all of the collection of times which correlate to events
+   * @param events A collection of all the events to get time ranges from
+   * @return A collection of times which correlate to events
+   */
+  private Collection<TimeRange> getTimeRangesFromEvents(Collection<Event> events) {
+    Collection<TimeRange> timeRanges = new ArrayList<TimeRange>();
+    for (Event event : events) {
+      timeRanges.add(event.getWhen());
+    }
+    return timeRanges;
+  }
+
+  /**
+   * Get a collection of times which are longer than or equal to in length to the {@code duration}
+   * and which do not overlap with the {@code occupiedTImes}
+   * @param occupiedTimes A collection of times which cannot overlap with the result
+   * @param duration The minimum length of time which can be returned in the collection
+   * @return A collection of times which are longer than or equal to in length to the {@code duration}
+   * and which do not overlap with the {@code occupiedTImes}
+   */
+  private Collection<TimeRange> getAvailableTimes(Collection<TimeRange> occupiedTimes,
+                                             long duration) {
+    if (occupiedTimes.size() == 0) {
+      if (TimeRange.WHOLE_DAY.duration() >= duration)
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      return Arrays.asList();
+    }
+
+    // order the times by the start
+    Collections.sort((ArrayList<TimeRange>) occupiedTimes, TimeRange.ORDER_BY_START);
```

Comment:
I believe the best way to change the Collection<TimeRange> to ArrayList<TimeRange> would be to either use the stream API or use new ArrayList constructor 
https://www.baeldung.com/java-convert-collection-arraylist

Reply:
Doing so would create a new list, to be sorted, should I work based off of that?

Reply:
Yeah I think that's preferred over typecasting